Sources/SwiftInspectorCore/InitializerAnalyzer.swift

```diff
@@ -0,0 +1,182 @@
+// Created by Francisco Diaz on 3/27/20.
+//
+// Copyright (c) 2020 Francisco Diaz
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import SwiftSyntax
+
+public final class InitializerAnalyzer: Analyzer {
+  /// - Parameter cachedSyntaxTree: The cached syntax tree to return the AST tree from
+  public init(name: String, cachedSyntaxTree: CachedSyntaxTree = .init()) {
+    self.name = name
+    self.cachedSyntaxTree = cachedSyntaxTree
+  }
+
+  public func analyze(fileURL: URL) throws -> [InitializerStatement] {
+    var statements: [InitializerStatement] = []
+    let syntax = try cachedSyntaxTree.syntaxTree(for: fileURL)
+
+    var reader = InitializerSyntaxReader(shouldVisitIdentifier: shouldVisit) { [unowned self] node in
+      let statement = self.initializerStatement(from: node)
+      statements.append(statement)
+    }
+    syntax.walk(&reader)
+
+    return statements
+  }
+
+  // MARK: Private
+  private let name: String
+  private let cachedSyntaxTree: CachedSyntaxTree
+
+  private func shouldVisit(_ identifier: String) -> Bool {
+    identifier == name
+  }
+
+  private func initializerStatement(from node: InitializerDeclSyntax) -> InitializerStatement {
+    let parameters = findParameters(from: node)
+    let modifiers = findModifiers(from: node)
+    return InitializerStatement(typeName: name, parameters: parameters, modifiers: modifiers)
+  }
+
+  private func findParameters(from node: InitializerDeclSyntax) -> [InitializerStatement.Parameter] {
+    let functionList = node.children
+      .compactMap { $0 as? ParameterClauseSyntax }
```

Comment:
Would it make sense to assert there is only one, to sanity check?

Reply:
I don't think anybody would be hitting that assertion. Remember this code is only used by running the binary or when we're developing so that assertion wouldn't be hit unless we add a test, or we can fail if that happens but that feels to harsh for the end-user.
If we think there's a use case for when we can have multiple `ParameterClauseSyntax` for one initializer I can add. a test but I couldn't come up with a use case.

Reply:
When we use this at work, do we build as Release or Debug?

At home I have an alias shell to run the debug version, so I could hit an assertion locally.

```shell
alias si="$HOME/repos/SwiftInspector/.build/debug/SwiftInspector"
``` 

I don't feel super strongly, though I see little downside in being generous with assertions for things that we truly think can never be hit. Even if you normally are using a Release build where they don't fire, you can switch to a Debug build if you are running into an unexpected issue. Then in that Debug build you have more insight into whether anything odd is happening, which can save debugging time.

And even if they never file, they can be a good way to signal developer intent. Reading the code right now it's not immediately clear if I should expect that there are multiple and we're picking the first, or that there should only ever be one. An assertion would be a way to communicate to a future maintainer that the latter is the case, and it's preferable to a comment IMHO since it does have meaning in the code (under some build configurations).

Anyway the length of this comment doesn't indicate that I feel super strongly. Mostly want to communicate my thinking 👍

Reply:
(Though I may not be using this to find initializer parameters at home, so I guess it's questionable if this would be hit 😂)